File: August 2025/Aug-11/.py
'''
Challenge Description:
---------------------
Given a string, determine whether the number of vowels in the first half of 
the string is equal to the number of vowels in the second half.

The string can contain any characters.
The letters a, e, i, o, and u, in either uppercase or lowercase, are considered vowels.
If there's an odd number of characters in the string, ignore the center character.
'''
import unittest

def is_balanced(s:str) -> bool:
    """
    Description:
    ------------
    Given a string, determine whether the number of vowels in the first half of 
    the string is equal to the number of vowels in the second half.
    
    Parameters:
    -----------
    - s: input string in which number of vowels in the 1st and 2nd halves are counted/compared

    Returns:
    --------
    - True if number of vowels in each half are eqaul
    - False if number of vowels in each half are NOT equal
    """
    vowels = ['a','e','i','o','u','A','E','I','O','U']
    first_half = 0
    sec_half = 0
    len_of_s = len(s)//2

    for char in range(len(s)): # For each character of the input string
        if s[char] in vowels:#   Check if that character is a vowel
            if char == len_of_s and len(s)%2 != 0:# If string is odd length and char is middle char, ignore
                continue
            if char < len_of_s:# If char is vowel and is in first half of string +1 to first half count
                first_half += 1
            elif char > len_of_s//2:# Else if char is vowel in 2nd half, +1 to second half count
                sec_half += 1

    return first_half == sec_half# return boolean answer to if vowels in 1st and 2nd half are equal

#================================ My thoughts on solution/improvement ================================
# Other possible solutions could include using RE library and using a regex
# to check if the character under review is a vowel
# could just uppercase the input  string and remove half the list of vowels
# This function runs in O(4n) time


#==================================== Solution from generative ai ====================================
def is_balanced(s: str) -> bool:
    vowels = set('aeiouAEIOU')
    mid = len(s) // 2

    first_half = sum(1 for c in s[:mid] if c in vowels)
    sec_half   = sum(1 for c in s[mid + (len(s) % 2):] if c in vowels)
    return first_half == sec_half


#==================================== Unit Tests ======================================================

class TestIsBalanced(unittest.TestCase):
    def test_racecar(self):
        self.assertEqual(is_balanced("racecar") , True)

    def test_latin(self):
        self.assertEqual(is_balanced("Lorem Ipsum") , True)
    
    def test_KittyIpsum(self):
        self.assertEqual(is_balanced("Kitty Ipsum") , False)

    def test_string(self):
        self.assertEqual(is_balanced("string") , False)

    def test_empty_string(self):
        self.assertEqual(is_balanced(" ") , True)

    def test_alphabet(self):
        self.assertEqual(is_balanced("abcdefghijklmnopqrstuvwxyz") , False)

    def test_special_char(self):
        self.assertEqual(is_balanced("123A#b!E&*456-0.U") , False)




if __name__ == "__main__":
    unittest.main(verbosity=2)

File: August 2025/Aug-11/.c
#include <stdio.h>      // Standard input and output library for terminal printing
#include <stdbool.h>    // Standard bool library to be abe to use boolean vallues
#include <string.h>     // Library that gives me ability to use strlen()

bool is_balanced(char s[]); // Function declaration (or prototype)

int main(){ // No comparable library to pythons unit-test, created my own tests
    printf("Test1 %s\n", is_balanced("racecar")     == true ? "Pass":"Fail");
    printf("Test2 %s\n", is_balanced("Lorem Ipsum") == true ? "Pass":"Fail");
    printf("Test3 %s\n", is_balanced("Kitty Ipsum") == false ? "Pass":"Fail");
    printf("Test4 %s\n", is_balanced("string")      == false ? "Pass":"Fail");
    printf("Test5 %s\n", is_balanced(" ")           == true ? "Pass":"Fail");
    printf("Test6 %s\n", is_balanced("abcdefghijklmnopqrstuvwxyz") == false ? "Pass":"Fail");
    printf("Test7 %s\n", is_balanced("123A#b!E&*456-0.U")           == false ? "Pass":"Fail");
    return 0;
}


bool is_balanced(char s[]){
    //Description:
    // ------------
    // Given a string, determine whether the number of vowels in the first half of 
    // the string is equal to the number of vowels in the second half.
    
    // Parameters:
    // -----------
    // - s: input string in which number of vowels in the 1st and 2nd halves are counted/compared

    // Returns:
    // --------
    // - True if number of vowels in each half are eqaul
    // - False if number of vowels in each half are NOT equal

    char vowels[] = "aeiouAEIOU";
    int first_half = 0;
    int sec_half = 0;
    int middle = strlen(s)/2;

    for(int i=0; i<strlen(s); i++){             // For each character in the input string
        for(int j=0; j<strlen(vowels); j++){    // for each character in the string of vowels
            if(vowels[j] == s[i]){              // If character from input is in string of vowels check the following
                if(i == middle && strlen(s)%2 != 0){ // if it is the middle character of an odd length input, ignore
                    continue;
                }
                if(i<middle){                   // If it is before the middle character increment first half count
                    first_half++;
                }
                else if(i > middle-1){          // If after middle character, increment 2nd half count
                    sec_half++;
                }
            }
        }
    }
    return first_half == sec_half; // Return boolean value
}

// My thoughts/ realizations
// 1. I could use int values of 0 and 1 instead of importing bool header file
// 2. Time complexity is O(n^2)
// 3. I could have used "if (strchr(vowels, s[i])) { ... }" to check if s[i] is in vowels (becomes O(n) complexity)
// 4. I am calling strlen() every iteration of the loop at that point make it a variable
// 5. I could have used Upper() string function to reduce the number of vowels in comparrison string
// 6. since input string is never changed making it a const could just help catch mistakes 







//   AI solution
#include <stdio.h>
#include <string.h>
#include <ctype.h>


int is_vowel(char c) {
    c = tolower(c);
    return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
}

int is_balanced(const char *s) {
    int len        = strlen(s);
    int half       = len / 2;
    int is_odd     = len % 2;
    int first_half = 0;
    int sec_half   = 0;

    for (int i = 0; i < len; i++) {
        if (!is_vowel(s[i])) continue;

        if (i == half && is_odd) continue;  // Skip middle char if odd length

        if (i < half)
            first_half++;
        else
            sec_half++;
    }

    return first_half == sec_half;
}

/*================================ Unit Tests ================================*/

typedef struct {
    const char *input;
    int         expected;
    const char *name;
} TestCase;

int run_tests() {
    TestCase tests[] = {
        { "racecar",                    1, "test_racecar"      },
        { "Lorem Ipsum",                1, "test_latin"        },
        { "Kitty Ipsum",                0, "test_KittyIpsum"   },
        { "string",                     0, "test_string"       },
        { " ",                          1, "test_empty_string" },
        { "abcdefghijklmnopqrstuvwxyz", 0, "test_alphabet"     },
        { "123A#b!E&*456-0.U",          0, "test_special_char" },
    };

    int num_tests = sizeof(tests) / sizeof(tests[0]);
    int passed    = 0;

    for (int i = 0; i < num_tests; i++) {
        int result = is_balanced(tests[i].input);
        int ok     = result == tests[i].expected;
        printf("%s (%s) ... %s\n",
               tests[i].name,
               tests[i].input,
               ok ? "OK" : "FAIL");
        if (ok) passed++;
    }

    printf("\n%d/%d tests passed.\n", passed, num_tests);
    return passed == num_tests ? 0 : 1;
}

int main() {
    return run_tests();
}